docs/tutorial/saxpy.rst

```diff
@@ -0,0 +1,780 @@
+file in this repository. The HIP code here mostly deals with where data
+has to be and when, and how devices transform this data. The first HIP calls
+deal with allocating device-side memory and dispatching data from host-side
+memory in a C runtime-like fashion.
+
+.. code-block:: C++
+
+  // Allocate and copy vectors to device memory.
+  float* d_x{};
+  float* d_y{};
+  HIP_CHECK(hipMalloc(&d_x, size_bytes));
+  HIP_CHECK(hipMalloc(&d_y, size_bytes));
+  HIP_CHECK(hipMemcpy(d_x, x.data(), size_bytes, hipMemcpyHostToDevice));
+  HIP_CHECK(hipMemcpy(d_y, y.data(), size_bytes, hipMemcpyHostToDevice));
+
+``HIP_CHECK`` is a custom macro borrowed from the examples utilities which
+checks the error code returned by API functions for errors and reports them to
+the console. It's not essential to the API.
+
+The code selects the device to allocate to and to dispatch to. Commands are
+issued to the HIP runtime per thread, and every thread has a device set as the
+target of commands. The default device is ``0``, which is equivalent to calling
+``hipSetDevice(0)``.
+
+Launch the calculation on the device after the data has been dispatched.
+
+.. code-block:: C++
+
+  __global__ void saxpy_kernel(const float a, const float* d_x, float* d_y, const unsigned int size)
+  {
+      // ...
+  }
+
+  int main()
+  {
+      // ...
+
+      // Launch the kernel on the default stream.
+      saxpy_kernel<<<dim3(grid_size), dim3(block_size), 0, hipStreamDefault>>>(a, d_x, d_y, size);
+  }
+
+Analyze at the signature of the offloaded function:
+
+- ``__global__`` instructs the compiler to generate code for this function as an
+  entrypoint to a device program, such that it can be launched from the host.
+- The function does not return anything, because there is no trivial way to
+  construct a return channel of a parallel invocation. Device-side entrypoints
+  may not return a value, their results should be communicated using output
+  parameters.
+- Device-side functions are typically called compute kernels, or just kernels
+  for short. This is to distinguish them from non-graphics-related graphics
+  shaders, or just shaders for short.
+- Arguments are taken by value and all arguments shall be
+  `TriviallyCopyable <https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable>`_,
+  meaning they should be `memcpy`-friendly. _(Imagine if they had custom copy
+  constructors. Where would that logic execute? On the host? On the device?)_
+  Pointer arguments are pointers to device memory, one typically backed by
+  VRAM.
+- We said that we'll be computing :math:`a\cdot x+y=z`, however we only pass
+  two pointers to the function. We'll be canonically reusing one of the inputs
+  as outputs.
+
+This function is launched from the host using a language extension often called
+the triple chevron syntax. Inside the angle brackets, provide the following.
+
+- The number of blocks to launch (our grid size)
```

Comment:
Do we have any documentation describing grid and block sizes? If yes can we point the user to it from here. I think these are non trivial concepts and user who just wants to exec stuff on their GPU can go it it if they want to.

Reply:
Added.